tsdb/store.go

```diff
@@ -1124,23 +1124,42 @@ func (s *Store) DeleteMeasurement(database, name string) error {
 	})
 }
 
-// filterShards returns a slice of shards where fn returns true
+// filterShards takes in a list of functions and
+// returns a slice of shards where fn returns true
 // for the shard. If the provided predicate is nil then all shards are returned.
 // filterShards should be called under a lock.
-func (s *Store) filterShards(fn func(sh *Shard) bool) []*Shard {
+func (s *Store) filterShards(fns []func(sh *Shard) bool) []*Shard {
 	var shards []*Shard
-	if fn == nil {
+
+	// Early return if nil or 0 len
+	if len(fns) == 0 || fns == nil {
 		shards = make([]*Shard, 0, len(s.shards))
-		fn = func(*Shard) bool { return true }
-	} else {
-		shards = make([]*Shard, 0)
+		fn := func(*Shard) bool { return true }
+
+		for _, sh := range s.shards {
+			if fn(sh) {
+				shards = append(shards, sh)
+			}
+		}
+
+		return shards
 	}
 
-	for _, sh := range s.shards {
-		if fn(sh) {
-			shards = append(shards, sh)
+	for _, fn := range fns {
+		if fn == nil {
+			shards = make([]*Shard, 0, len(s.shards))
+			fn = func(*Shard) bool { return true }
+		} else {
+			shards = make([]*Shard, 0)
+		}
+
+		for _, sh := range s.shards {
+			if fn(sh) {
+				shards = append(shards, sh)
+			}
```

Comment:
Don't we end up throwing away `shards` when we move onto the next `fn`?

Reply:
Yes - see comment in reply to David's 👍 

I need to adjust this, definitely a bug 😅